client/fingerprint_manager.go

```diff
@@ -0,0 +1,122 @@
+package client
+
+import (
+	"log"
+	"time"
+
+	"github.com/hashicorp/nomad/client/config"
+	"github.com/hashicorp/nomad/client/driver"
+	"github.com/hashicorp/nomad/client/fingerprint"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// FingerprintManager runs a client fingerprinters on a continuous basis, and
+// updates the client when the node has changed
+type FingerprintManager struct {
+	getConfig func() *config.Config
+	node      *structs.Node
+	client    *Client
+	logger    *log.Logger
+}
+
+// run runs each fingerprinter individually on an ongoing basis
+func (fm *FingerprintManager) run(f fingerprint.Fingerprint, period time.Duration, name string) {
+	fm.logger.Printf("[DEBUG] fingerprint_manager: fingerprinting %s every %v", name, period)
+
+	for {
+		select {
+		case <-time.After(period):
+			_, err := fm.fingerprint(name, f)
+			if err != nil {
+				fm.logger.Printf("[DEBUG] fingerprint_manager: periodic fingerprinting for %v failed: %+v", name, err)
+				continue
+			}
+
+		case <-fm.client.shutdownCh:
+			return
+		}
+	}
+}
+
+// setupDrivers is used to fingerprint the node to see if these drivers are
+// supported
+func (fm *FingerprintManager) SetupDrivers(drivers []string) error {
+	var availDrivers []string
+	driverCtx := driver.NewDriverContext("", "", fm.getConfig(), fm.node, fm.logger, nil)
+	for _, name := range drivers {
+
+		d, err := driver.NewDriver(name, driverCtx)
+		if err != nil {
+			return err
```

Comment:
Question about error handling during loop iteration - if some drivers  worked and one returned an error in line 49, this returns immediatly. It could have already started periodic fingerprinting some drivers before it failed and returned. Is that desirable?

Reply:
This should exit fast if there is an error during setting up the client. Otherwise, it should gracefully handle errors during periodic fingerprinting.